sqlx-sqlite/src/connection/mod.rs

```diff
@@ -172,6 +212,21 @@ impl Connection for SqliteConnection {
     }
 }
 
+/// Implements a C binding to a progress callback. The function returns `0` if the
+/// user-provided callback returns `true`, and `1` otherwise to signal an interrupt.
+extern "C" fn progress_callback<F>(callback: *mut c_void) -> c_int
+where
+    F: FnMut() -> bool,
+{
+    unsafe {
+        if (*(callback as *mut F))() {
```

Comment:
:stop_sign: This is undefined behavior in the case that the callback panics, as that would unwind across the FFI boundary.

Reply:
You can use [`std::panic::catch_unwind()`](https://doc.rust-lang.org/stable/std/panic/fn.catch_unwind.html) to halt unwinding, but then we have to decide what to _do_ with the panic.

If we want to just log the panic and move on, we should probably require [`F: UnwindSafe`](https://doc.rust-lang.org/stable/std/panic/trait.UnwindSafe.html), and perhaps use `Fn()` instead of `FnMut()`.

Alternatively, we could [abort](https://doc.rust-lang.org/stable/std/process/fn.abort.html), but that could lead to some _very_ unhappy users.

Reply:
We could store the panic somewhere until control flow returns to our code, then [propagate it](https://doc.rust-lang.org/stable/std/panic/fn.resume_unwind.html), but that would just crash the worker thread anyway.